File: crates/breez-sdk/bindings/examples/cli/langs/csharp/Webhooks.cs
using Breez.Sdk.Spark;

namespace BreezCli;

/// <summary>
/// Represents a single webhooks subcommand.
/// </summary>
public class WebhookCliCommand
{
    public required string Name { get; init; }
    public required string Description { get; init; }
    public required Func<BreezSdk, Func<string, string?>, string[], Task> Run { get; init; }
}

/// <summary>
/// Webhook subcommand names (used for REPL completion).
/// </summary>
public static class WebhookCommandNames
{
    public static readonly string[] All =
    {
        "webhooks register",
        "webhooks unregister",
        "webhooks list",
    };
}

/// <summary>
/// Webhook subcommand handlers.
/// </summary>
public static class WebhookCommands
{
    /// <summary>
    /// Builds the webhooks subcommand registry.
    /// </summary>
    public static Dictionary<string, WebhookCliCommand> BuildRegistry()
    {
        return new Dictionary<string, WebhookCliCommand>
        {
            ["register"] = new()
            {
                Name = "register",
                Description = "Register a new webhook",
                Run = HandleRegister
            },
            ["unregister"] = new()
            {
                Name = "unregister",
                Description = "Unregister a webhook",
                Run = HandleUnregister
            },
            ["list"] = new()
            {
                Name = "list",
                Description = "List all registered webhooks",
                Run = HandleList
            },
        };
    }

    /// <summary>
    /// Dispatches a webhooks subcommand.
    /// </summary>
    public static async Task DispatchCommand(
        string[] args,
        BreezSdk sdk,
        Func<string, string?> readline)
    {
        var registry = BuildRegistry();

        if (args.Length == 0 || args[0] == "help")
        {
            Console.WriteLine();
            Console.WriteLine("Webhook subcommands:");
            var names = registry.Keys.OrderBy(k => k).ToList();
            foreach (var name in names)
            {
                Console.WriteLine($"  webhooks {name,-30} {registry[name].Description}");
            }
            Console.WriteLine();
            return;
        }

        var subName = args[0];
        var subArgs = args.Skip(1).ToArray();

        if (!registry.TryGetValue(subName, out var cmd))
        {
            Console.WriteLine($"Unknown webhooks subcommand: {subName}. Use 'webhooks help' for available commands.");
            return;
        }

        await cmd.Run(sdk, readline, subArgs);
    }

    // -----------------------------------------------------------------------
    // Webhook command handlers
    // -----------------------------------------------------------------------

    private static WebhookEventType? ParseEventType(string s)
    {
        return s.ToLower() switch
        {
            "lightning-receive" => new WebhookEventType.LightningReceiveFinished(),
            "lightning-send" => new WebhookEventType.LightningSendFinished(),
            "coop-exit" => new WebhookEventType.CoopExitFinished(),
            "static-deposit" => new WebhookEventType.StaticDepositFinished(),
            _ => null
        };
    }

    // --- register ---

    private static async Task HandleRegister(BreezSdk sdk, Func<string, string?> readline, string[] args)
    {
        if (args.Length < 3)
        {
            Console.WriteLine("Usage: webhooks register <url> <secret> <event_type> [<event_type> ...]");
            Console.WriteLine("Event types: lightning-receive, lightning-send, coop-exit, static-deposit");
            return;
        }

        var url = args[0];
        var secret = args[1];
        var eventTypes = new List<WebhookEventType>();

        for (int i = 2; i < args.Length; i++)
        {
            var eventType = ParseEventType(args[i]);
            if (eventType == null)
            {
                Console.WriteLine($"Unknown event type: {args[i]}. Valid values: lightning-receive, lightning-send, coop-exit, static-deposit");
                return;
            }
            eventTypes.Add(eventType);
        }

        var result = await sdk.RegisterWebhook(request: new RegisterWebhookRequest(
            url: url,
            secret: secret,
            eventTypes: eventTypes.ToArray()
        ));
        Serialization.PrintValue(result);
    }

    // --- unregister ---

    private static async Task HandleUnregister(BreezSdk sdk, Func<string, string?> readline, string[] args)
    {
        if (args.Length < 1)
        {
            Console.WriteLine("Usage: webhooks unregister <webhook_id>");
            return;
        }

        await sdk.UnregisterWebhook(request: new UnregisterWebhookRequest(webhookId: args[0]));
        Console.WriteLine("Webhook unregistered successfully");
    }

    // --- list ---

    private static async Task HandleList(BreezSdk sdk, Func<string, string?> readline, string[] args)
    {
        var webhooks = await sdk.ListWebhooks();
        Serialization.PrintValue(webhooks);
    }
}

File: crates/breez-sdk/bindings/examples/cli/langs/golang/webhooks.go
package main

import (
	"fmt"
	"sort"
	"strings"

	breez_sdk_spark "github.com/breez/breez-sdk-spark-go/breez_sdk_spark"
	"github.com/chzyer/readline"
)

// WebhookCmd represents a single webhooks subcommand.
type WebhookCmd struct {
	Name        string
	Description string
	Run         func(sdk *breez_sdk_spark.BreezSdk, rl *readline.Instance, args []string) error
}

// WebhookCommandNames lists all webhooks subcommand names (used for REPL completion).
var WebhookCommandNames = []string{
	"webhooks register",
	"webhooks unregister",
	"webhooks list",
}

// BuildWebhookRegistry returns a map of webhooks subcommand name to WebhookCmd.
func BuildWebhookRegistry() map[string]WebhookCmd {
	return map[string]WebhookCmd{
		"register": {
			Name:        "register",
			Description: "Register a new webhook",
			Run:         handleWebhookRegister,
		},
		"unregister": {
			Name:        "unregister",
			Description: "Unregister a webhook",
			Run:         handleWebhookUnregister,
		},
		"list": {
			Name:        "list",
			Description: "List all registered webhooks",
			Run:         handleWebhookList,
		},
	}
}

// DispatchWebhookCommand dispatches a webhooks subcommand.
func DispatchWebhookCommand(args []string, sdk *breez_sdk_spark.BreezSdk, rl *readline.Instance) {
	registry := BuildWebhookRegistry()

	if len(args) == 0 || args[0] == "help" {
		fmt.Println("\nWebhooks subcommands:")
		names := make([]string, 0, len(registry))
		for name := range registry {
			names = append(names, name)
		}
		sort.Strings(names)
		for _, name := range names {
			cmd := registry[name]
			fmt.Printf("  webhooks %-30s %s\n", name, cmd.Description)
		}
		fmt.Println()
		return
	}

	subName := args[0]
	subArgs := args[1:]

	cmd, ok := registry[subName]
	if !ok {
		fmt.Printf("Unknown webhooks subcommand: %s. Use 'webhooks help' for available commands.\n", subName)
		return
	}

	if err := cmd.Run(sdk, rl, subArgs); err != nil {
		fmt.Printf("Error: %v\n", err)
	}
}

// ---------------------------------------------------------------------------
// Webhooks command handlers
// ---------------------------------------------------------------------------

// --- register ---

func handleWebhookRegister(sdk *breez_sdk_spark.BreezSdk, _ *readline.Instance, args []string) error {
	if len(args) < 3 {
		fmt.Println("Usage: webhooks register <url> <secret> <event_type> [<event_type> ...]")
		fmt.Println("Event types: lightning-receive, lightning-send, coop-exit, static-deposit")
		return nil
	}

	url := args[0]
	secret := args[1]
	eventTypes, err := parseWebhookEventTypes(args[2:])
	if err != nil {
		return err
	}

	result, err := sdk.RegisterWebhook(breez_sdk_spark.RegisterWebhookRequest{
		Url:        url,
		Secret:     secret,
		EventTypes: eventTypes,
	})
	if err = liftError(err); err != nil {
		return err
	}
	printValue(result)
	return nil
}

// --- unregister ---

func handleWebhookUnregister(sdk *breez_sdk_spark.BreezSdk, _ *readline.Instance, args []string) error {
	if len(args) < 1 {
		fmt.Println("Usage: webhooks unregister <webhook_id>")
		return nil
	}

	if err := liftError(sdk.UnregisterWebhook(breez_sdk_spark.UnregisterWebhookRequest{
		WebhookId: args[0],
	})); err != nil {
		return err
	}
	fmt.Println("Webhook unregistered successfully")
	return nil
}

// --- list ---

func handleWebhookList(sdk *breez_sdk_spark.BreezSdk, _ *readline.Instance, _ []string) error {
	result, err := sdk.ListWebhooks()
	if err = liftError(err); err != nil {
		return err
	}
	printValue(result)
	return nil
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

func parseWebhookEventTypes(values []string) ([]breez_sdk_spark.WebhookEventType, error) {
	var types []breez_sdk_spark.WebhookEventType
	for _, v := range values {
		switch strings.ToLower(v) {
		case "lightning-receive":
			types = append(types, breez_sdk_spark.WebhookEventTypeLightningReceiveFinished{})
		case "lightning-send":
			types = append(types, breez_sdk_spark.WebhookEventTypeLightningSendFinished{})
		case "coop-exit":
			types = append(types, breez_sdk_spark.WebhookEventTypeCoopExitFinished{})
		case "static-deposit":
			types = append(types, breez_sdk_spark.WebhookEventTypeStaticDepositFinished{})
		default:
			return nil, fmt.Errorf("unknown webhook event type: %s (valid: lightning-receive, lightning-send, coop-exit, static-deposit)", v)
		}
	}
	return types, nil
}
